### PR TITLE
 FSE: Add Current layout to starter page templates page layout picker when editing a homepage

### PR DESCRIPTION
**Pull request by andrewserong**
#### Changes proposed in this Pull Request

* Add Current layout option to page layout picker when a user is editing / updating a homepage

#### Still to do

* [ ] Switch on dynamic previews for _just_ the Current button, and see if that works instead of having a screenshot or empty square for this button
* [x] The current option shouldn't display if this is a brand new page

#### Screenshot

image

#### Testing instructions

* Spin up a local development environment (e.g. go to the `apps/full-site-editing` directory and run `yarn wp-env start`
* Run the local dev environment for FSE (e.g. fo to the `apps/full-site-editing` directory and run `yarn dev` or if you want to sync to your sandbox, run `yarn dev --sync`)
* In your WP install, set a page to be the front page in the customizer
* Go to edit that page, and then add the `?new-homepage` query param so that you open up the page layout selector for a homepage, where you should then see the Current layout option. E.g. for my own local test site, the URL is `http://localhost:4013/wp-admin/post.php?post=2&action=edit&new-homepage`
* With Current layout selected, you should see a preview of the current blocks in your page, and the page title should be correct in the preview
* Try selecting different page layouts — these should all replace the page's content, however if you select Current layout, it should _not_ replace the page content.

Related to: #43933

### CODE REVIEW

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (a8ci18n) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:430]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **100** times:
`translate( 'Current' )` *ES Score: 10*

**Comment by andrewserong** [comment]
Nice work on this @Addison-Stavlo — I've done a small update to move the blocks for the `isCurrentPreview` to within the `renderTemplatesList` function. Because `getBlocksByTemplateSlugs` is memoized, it was returning no blocks for `current` if the selector hadn't finished resolving before it was first called. This meant that I was intermittently seeing there was no "Current" layout option because for some reason in my local environment, it was taking a little longer to resolve that selector. This change isn't quite as neat, but means that when the selector updates, the template list should display correctly.

I looked into the `getBoundingClientRect` issue and it looks or _feels_ like there's a Popover somewhere that's trying re-adjust its size, and I'm wondering if somehow the block provider expects there to be a Popover.Slot — or because our parent document's block provider has one, it's somehow expecting something like that here? I found a workaround by adding `tab-index=-1` to the iframe, to prevent the user from tabbing into it, which seems to suppress that error, and enable tabbing past the Current layout option. I can see the issue is still present in the main preview if you click into it and tab around, but that's separate from our proposed changes here.

So, it looks like we might be okay without having to add back in `Disabled`?

**Comment by andrewserong** [comment]
Oh, and yes, I agree, if we get this working acceptably, I think it’d be great to include the current layout option for when a page already exists and someone is looking to swap layouts.

Another thought re: the getBoundingClientRect error being thrown while tabbing within a preview — it’s almost like it’s trying to open the toolbar pop up for each block or something like that? More broadly it feels like this approach to previews has a lot more going on than we might need to a non-interactive preview, but I see it’s very similar to the block pattern previews in Gutenberg. Just a bit of food for thought if we get the time to rethink these previews in the future!

**Comment by Addison-Stavlo** [comment]
Very nice, the `tabIndex` there solves it. 😅 I was trying tabIndex in a couple places but I must have goofed on trying the iframe element itself.  Its odd how that doesn't solve the same issue for the large preview though 🤔  although thats less of a concern for this PR I think.

**Comment by Addison-Stavlo** [comment]
Testing this some more it seems to work well!  Especially considering that the error we have encountered seems to be present on master and is a recent problem stemming from core Gutenberg after 8.3.

I removed the check for `isFrontPage` so the current content option should be available for all pages.  Il set this to 'ready for review' (I hope thats ok), but it seems we will need to approve this and deploy an FSE plugin release for it before we can merge  #43933

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (a8ci18n) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:443]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **100** times:
`translate( 'Current' )` *ES Score: 10*

**Comment by alshakero** [comment]
Hi Andy and Addison, 

> The current option shouldn't display if this is a brand-new page

I implemented this one, although I'm not entirely sure about the validity of the criteria I used to tell if the page is new. I took inspiration from [this line](https://github.com/Automattic/wp-calypso/blob/update/fse-starter-page-templates-to-include-current-layout/apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js#L173). And it seems to work as expected.

As for
> Switch on dynamic previews for just the Current button, and see if that works instead of having a screenshot or empty square for this button

This seems to be already done. I changed the text color to red and the button reflected that alright:

image

What's not done though is that the big preview isn't showing the "current". 


What was off though is the big preview window, it didn't show "Current". I changed `getDefaultSelectedTemplate` to select `Current` when the page isn't new. I'm not entirely confident about this change though. Would appreciate some feedback.

Thanks!

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (a8ci18n) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:176]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **100** times:
`translate( 'Current' )` *ES Score: 10*

File: apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js
Comment (Addison-Stavlo) [inline apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:353]: Similarly, this is what I had added as the condition to hide the 'Current' section on a new page.  I was assuming that a new page would always have no content, or that if it did have content we wouldn't want to hide 'Current' ? 🤔   Similarly if it is not a new page and has no content, we don't need the 'current' section.

Comment (alshakero) [reply apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js:353]: I think these are valid assumptions. We probably don't need "Current" for any empty page. New or not.

**Comment by alshakero** [comment]
> My apologies as I forgot to update the original comment to reflect the state of these TODOs.

No worries. I probably should have paid more attention to the convo here. I reversed my redundant check and only kept the change that renders the "current" contents in the preview iframe. 

I tested and all works as intended.

**Comment by ockham** [comment]
Going to rebase.

**Comment by Copons** [comment]
Rebased again as there was an issue in an unrelated sub-plugin that were crashing wp-admin. 🙂

**Comment by Copons** [comment]
This looked good to me and tested fine...
But then I _really_ had to go in and replace an unnecessary `let` with a `const`, so now it feels awkward to self-approve the PR. 😅 

I'll leave it around for a bit longer in case someone has something to add, otherwise I think this is ready to merge. ✨

**Comment by andrewserong** [comment]
@Copons I believe that was the outstanding issue, thanks for picking this one up! Next up is to test that it works okay across browsers / mobile devices, that the tabbing behaviour is working okay, and not doing anything strange (e.g. https://github.com/Automattic/wp-calypso/issues/44038), and check that the preview performance is acceptable.

And it'll need _another_ rebase once https://github.com/Automattic/wp-calypso/pull/44501 lands, too of course.

**Comment by Copons** [comment]
Rebased after #44501

**Comment by Addison-Stavlo** [comment]
> But then I really had to go in and replace an unnecessary `let` with a `const`

You couldn't just `let` it be?   :trollface:  🤣   

> so now it feels awkward to self-approve the PR

lol, yeah it feels awkward for me to self-approve as well since I worked on it some previously as well.

✅   Consider your changes in c3df4b0537a8a77db2fe861303e706556f95a071 approved via this comment if it makes approving the rest feel less awkward. 😁 

> Is there anything else that's missing?

I can't think of anything beyond what Andrew just said above.  Thanks for picking this up!

File: apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/index.js
Comment (Copons) [inline apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/index.js:177]: I've got a doubt about this logic here:

```js
if ( ! props.isFrontPage && ! previouslyChosenTemplate ) {
	return blankTemplate;
}
return 'current';
```

As it is written right now, the Current template can only be autoselected when editing the front page.
Is this intentional?
Do we have a reason for checking `previouslyChosenTemplate` if we then end up selecting either Blank or Current (which is a "special" case)?

What about autoselecting Current for _any_ page, as long as they have `previouslyChosenTemplate`?
E.g.
```js
if ( props.isFrontPage || previouslyChosenTemplate ) {
	return 'current';
}
return blankTemplate;
```

It would mean that:
- Front page always autoselect Current. (This is my doubt)
- Other pages autoselect Current if they are using a template, or Blank otherwise (e.g. new pages).

**Comment by Copons** [comment]
Other than a [relatively minor doubt](https://github.com/Automattic/wp-calypso/pull/43980#pullrequestreview-475632242), I'm seeing two visual glitches:

- When the Current template is autoselected, the post title doesn't resize appropriately:

Screenshot 2020-08-26 at 16 43 29

This is caused by the `PostTitle` component failing to resize when [it obtains the value via DOM manipulation](https://github.com/Automattic/wp-calypso/blob/d4fcffe744c8868bbf085d0eb80e7696c3a65986/apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/components/block-iframe-preview.js#L128-L153).
The code there mentions a [Core PR](https://github.com/WordPress/gutenberg/pull/20609/) that will enable us to set the title directly via props instead.
It seems it would fix this issue, but meanwhile we are stuck with the DOM manipulation.

One way we could fix this is to replace [`<PostTitle />`](https://github.com/Automattic/wp-calypso/blob/d4fcffe744c8868bbf085d0eb80e7696c3a65986/apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/components/block-preview.js#L28) with the actual rendered code of the component.
```jsx
<div class="wp-block editor-post-title editor-post-title__block">
	<div class="editor-post-title__input">{ title }</div>
</div>
```
It is slightly fragile, but we might want to assume that the `PostTitle` rendered content won't change much before https://github.com/WordPress/gutenberg/pull/20609/ is merged. 🤔 

---

- When opening the template selector with Current autoselected, there's a very very brief FOUC.

Screenshot 2020-08-26 at 17 06 18

It's likely caused by [`copyStylesToIframe`](https://github.com/Automattic/wp-calypso/blob/fdcd9427bac872b559c01c650ce75f4030071454/apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/components/block-iframe-preview.js#L158) taking more time than the preview to be populated with blocks.

Given the flash is barely noticeable, and the fix could be complicated, we should probably rather tackle it in a follow up.

File: apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/index.js
Comment (Addison-Stavlo) [reply apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/index.js:177]: I think these two logic expressions are mathematically equivalent via Demorgan?  It seems like they should have the same effect, although the second seems a bit more readable.

We have basically gone from: ` !A && !B => do C `. to ` ! (A || B) => do C` to `A || B => do !C` ? 🤷‍♀️   I think it will be a more readable change but it doesn't seem to cause any difference in functionality?

**Comment by Addison-Stavlo** [comment]
> One way we could fix this is to replace <PostTitle /> with the actual rendered code of the component.

I think that seems reasonable for the time being?  the DOM manipulation is fragile as-is, I don't think this is going to make it much more fragile in comparison?  It will be great when we can get rid of the DOM manipulation all together.

> Given the flash is barely noticeable, and the fix could be complicated, we should probably rather tackle it in a follow up.

That sounds fair as well.  Although are there circumstances where this would take much longer and be very noticeable? 🤔

File: apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/index.js
Comment (Copons) [reply apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-modal/index.js:177]: Right, I think I was testing with some very dirty data, and so my change gave me some kind of false positive. 😕 
I'll leave this alone.

**Comment by Copons** [comment]
> That sounds fair as well. Although are there circumstances where this would take much longer and be very noticeable? 🤔

It's definitely noticeable on slow connections.
I'd argue that slow connections are already used to FOUCs, to be honest.

---

I've also given more thought about replacing `PostTitle` with its rendered content, and I think that too should be tackled separately.
The title handling affects various parts of the modal, and removing the component means also deleting a good amount of code, which can lead to regressions and whatnot.

Given that this PR works, and the 2 updates are normal (the title cut off in some cases) or low (the FOUC) priority, I'd say let's merge this and try to tackle the updates before the next Editing Toolkit release.

**Comment by Addison-Stavlo** [comment]
>  I'd say let's merge this and try to tackle the updates before the next Editing Toolkit release.

That sounds fair.

Other possible ideas for the title issue - Either remove it from from the preview when it is showing 'current' or maybe just hardcode it as 'Current Layout'.  When you select other layouts to look at previews, the title it shows ends up being title of the preview and not the actual post title anyways? 🤔

**Comment by david-szabo97** [comment]
Can't we hack the **PostTitle** for now? I know it's very hacky but might be a good temporary solution to add a style to override the height of the post title.
```js
iframeRef.current.contentDocument.head.innerHTML += '<style>.editor-post-title .editor-post-title__input { height: auto !important; }</style>';
```

The current layout never shows up in the preview. Is it just happening for me? 😕 
image
I noticed the `currentBlocks` is an empty array `[]`. Since we `memoize` the `getBlocksForPreview` function, it always stay an empty array. A quick hack would be to clear the cache if the `currentBlocks` is loaded:
```js
const currentBlocksPreviewCache = this.getBlocksForPreview.cache.get( 'current' );
if ( currentBlocksPreviewCache && currentBlocksPreviewCache.length !== currentBlocks.length ) {
	this.getBlocksForPreview.cache.delete( 'current' );
}
```

**Comment by Addison-Stavlo** [comment]
> The current layout never shows up in the preview. Is it just happening for me?

The current layout seems to show up for me, I wonder what the difference here is 🤔 .  Either way, if clearing the cache fixes it on your end that seems reasonable to me.

> Can't we hack the PostTitle for now? 

Yeah, that seems fair.  Injecting the style like that should fix it without introducing any of the negatives @Copons was mentioning above.

**Review by Addison-Stavlo — APPROVED** [review]
Tested with the most recent changes and it works well on my end.  The current layout is displayed and the title issue is resolved.

**Pull request merged**
